File: apps/backend/app/services/merchant_normalizer.py
# apps/backend/app/services/merchant_normalizer.py

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Optional, Literal, TYPE_CHECKING

if TYPE_CHECKING:
    from redis.asyncio import Redis

MerchantKind = Literal["p2p", "subscription", "retail", "atm", "cash", "other"]
MerchantCategoryHint = Literal[
    "transfers",
    "subscriptions",
    "groceries",
    "fuel",
    "dining",
    "atm",
    "cash",
    "other",
    "unknown",
]


@dataclass
class NormalizedMerchant:
    display: str
    kind: Optional[MerchantKind] = None
    category_hint: Optional[MerchantCategoryHint] = None
    rule_id: Optional[str] = None  # for debugging


@dataclass
class MerchantBrandRule:
    id: str
    pattern: re.Pattern
    normalized: str
    kind: Optional[MerchantKind] = None
    category_hint: Optional[MerchantCategoryHint] = None


BRAND_RULES: list[MerchantBrandRule] = [
    MerchantBrandRule(
        id="zelle_now_withdrawal",
        pattern=re.compile(
            r"\b(now\s+withdrawal|zelle(?:\s+(payment|transfer))?)\b", re.I
        ),
        normalized="Zelle transfer",
        kind="p2p",
        category_hint="transfers",
    ),
    MerchantBrandRule(
        id="venmo",
        pattern=re.compile(r"\bvenmo\b", re.I),
        normalized="Venmo",
        kind="p2p",
        category_hint="transfers",
    ),
    MerchantBrandRule(
        id="cash_app",
        pattern=re.compile(r"\b(sq\s*\*|sqc\*|cash\s*app)\b", re.I),
        normalized="Cash App",
        kind="p2p",
        category_hint="transfers",
    ),
    MerchantBrandRule(
        id="paypal_p2p",
        pattern=re.compile(
            r"\bpaypal\b(?!.*\b(netflix|spotify|amazon|adobe|microsoft|apple)\b)", re.I
        ),
        normalized="PayPal",
        kind="p2p",
        category_hint="transfers",
    ),
    MerchantBrandRule(
        id="apple_cash",
        pattern=re.compile(r"\bapple\s*cash\b", re.I),
        normalized="Apple Cash",
        kind="p2p",
        category_hint="transfers",
    ),
]


def _basic_normalize(raw: str) -> str:
    if not raw:
        return "Unknown"

    s = raw.strip()

    # Simplify separators
    s = re.sub(r"[*_]+", " ", s)
    s = re.sub(r"\s{2,}", " ", s)

    # Drop long numeric / phone-like tails
    s = re.sub(
        r"\s+(x?\d{4,}|\d{3}-\d{3,}-\d{3,}|\d{10,}|[0-9]{4,}[A-Z0-9-]*)\s*$",
        "",
        s,
        flags=re.I,
    ).strip()

    if not s:
        return "Unknown"

    parts = s.lower().split()
    return " ".join(p.capitalize() for p in parts)


def normalize_merchant_for_category(raw: str) -> NormalizedMerchant:
    """
    Normalize a raw merchant/description string into structured data.

    This is the synchronous baseline normalizer. For async request handlers,
    prefer normalize_merchant_with_memory() which checks Redis first.

    Returns:
        NormalizedMerchant with display name, kind, and category hint
    """
    if not raw:
        return NormalizedMerchant(display="Unknown", category_hint="unknown")

    for rule in BRAND_RULES:
        if rule.pattern.search(raw):
            return NormalizedMerchant(
                display=rule.normalized,
                kind=rule.kind,
                category_hint=rule.category_hint or "other",
                rule_id=rule.id,
            )

    return NormalizedMerchant(
        display=_basic_normalize(raw),
        category_hint="unknown",
    )


async def normalize_merchant_with_memory(
    raw: Optional[str],
    redis: Optional[Redis] = None,
) -> NormalizedMerchant:
    """
    Preferred entrypoint for async request handlers.

    Strategy:
    1. Try Redis memory (user / ML / rule decisions from past 30 days)
    2. Fall back to regex + heuristics
    3. Store result back to Redis with 30d TTL

    This ensures consistency: once we've seen "NOW Withdrawal Zelle...",
    we always normalize it the same way without re-running regex.

    Args:
        raw: Raw merchant/description string from bank statement
        redis: Redis connection (optional, degrades gracefully if None)

    Returns:
        NormalizedMerchant with display name, kind, and category hint
    """
    if raw is None:
        return normalize_merchant_for_category("")

    raw_str = raw.strip()
    if not raw_str:
        return normalize_merchant_for_category("")

    # Try Redis memory first
    if redis is not None:
        try:
            from app.services.merchant_memory import (
                get_merchant_memory,
                put_merchant_memory,
            )

            mem = await get_merchant_memory(redis, raw_str)
            if mem is not None:
                return NormalizedMerchant(
                    display=mem.canonical,
                    kind=mem.kind,  # type: ignore
                    category_hint=mem.category_hint,  # type: ignore
                    rule_id=None,
                )
        except Exception:
            # Redis down → continue with heuristics
            pass

    # Fallback to synchronous normalizer
    norm = normalize_merchant_for_category(raw_str)

    # Store in Redis for future lookups
    if redis is not None:
        try:
            from app.services.merchant_memory import put_merchant_memory

            # Confidence 0.9 for rules, 0.7 for generic heuristic
            confidence = 0.9 if norm.rule_id else 0.7
            source = "rule" if norm.rule_id else "heuristic"
            await put_merchant_memory(
                redis, raw_str, norm, confidence=confidence, source=source
            )
        except Exception:
            # Redis unavailable → log but don't fail
            pass

    return norm

File: apps/web/src/lib/merchant-normalizer.ts
// apps/web/src/lib/merchant-normalizer.ts

export type MerchantKind =
  | "p2p"
  | "subscription"
  | "retail"
  | "atm"
  | "cash"
  | "other";

export type MerchantCategoryHint =
  | "transfers"
  | "subscriptions"
  | "groceries"
  | "fuel"
  | "dining"
  | "atm"
  | "cash"
  | "other"
  | "unknown";

export interface MerchantBrandRule {
  /** Stable id for debugging / tests */
  id: string;
  /** Pattern to match raw statement description or merchant field */
  pattern: RegExp;
  /** Normalized, user-facing merchant label */
  normalized: string;
  kind?: MerchantKind;
  categoryHint?: MerchantCategoryHint;
}

export interface NormalizedMerchant {
  /** Human readable label (e.g. "Zelle transfer") */
  display: string;
  /** Optional semantic kind (p2p, subscription, etc.) */
  kind?: MerchantKind;
  /** Optional category hint (e.g. "transfers") */
  categoryHint?: MerchantCategoryHint;
}

/**
 * Brand-specific rules that override generic cleanup.
 * Keep this small and high-signal – we only add brands we care about.
 */
export const MERCHANT_BRAND_RULES: MerchantBrandRule[] = [
  {
    id: "zelle_now_withdrawal",
    pattern: /\b(now\s+withdrawal|zelle(?:\s+(payment|transfer))?)\b/i,
    normalized: "Zelle transfer",
    kind: "p2p",
    categoryHint: "transfers",
  },
  {
    id: "venmo",
    pattern: /\bvenmo\b/i,
    normalized: "Venmo",
    kind: "p2p",
    categoryHint: "transfers",
  },
  {
    id: "cash_app",
    // Square Cash descriptors often start with "SQ *" or "SQC*"
    pattern: /\b(sq\s*\*|sqc\*|cash\s*app)\b/i,
    normalized: "Cash App",
    kind: "p2p",
    categoryHint: "transfers",
  },
  {
    id: "paypal_p2p",
    // Treat generic PayPal as P2P except when obviously a merchant name follows.
    pattern:
      /\bpaypal\b(?!.*\b(netflix|spotify|amazon|adobe|microsoft|apple)\b)/i,
    normalized: "PayPal",
    kind: "p2p",
    categoryHint: "transfers",
  },
  // Additional brand rules
  {
    id: "apple_cash",
    pattern: /\bapple\s*cash\b/i,
    normalized: "Apple Cash",
    kind: "p2p",
    categoryHint: "transfers",
  },
  {
    id: "playstation",
    pattern: /\b(playstatio|playstation)\b/i,
    normalized: "PlayStation",
    kind: "subscription",
    categoryHint: "subscriptions",
  },
  {
    id: "netflix",
    pattern: /\bnetflix\b/i,
    normalized: "Netflix",
    kind: "subscription",
    categoryHint: "subscriptions",
  },
  {
    id: "spotify",
    pattern: /\bspotify\b/i,
    normalized: "Spotify",
    kind: "subscription",
    categoryHint: "subscriptions",
  },
  {
    id: "starbucks",
    pattern: /\bstarbucks\b/i,
    normalized: "Starbucks",
    kind: "retail",
    categoryHint: "dining",
  },
  {
    id: "uber",
    pattern: /\buber(?!\s*eats)\b/i,
    normalized: "Uber",
    kind: "retail",
    categoryHint: "other",
  },
  {
    id: "lyft",
    pattern: /\blyft\b/i,
    normalized: "Lyft",
    kind: "retail",
    categoryHint: "other",
  },
  {
    id: "harris_teeter",
    pattern: /\bharris\s*teeter\b/i,
    normalized: "Harris Teeter",
    kind: "retail",
    categoryHint: "groceries",
  },
];

/**
 * Generic cleanup: strip noise, collapse spaces, remove long numeric tails.
 */
function basicNormalize(raw: string): string {
  if (!raw) return "Unknown";

  let s = raw.trim();

  // Normalize separators a bit
  s = s.replace(/[*_]+/g, " ");
  s = s.replace(/\s{2,}/g, " ");

  // Drop long trailing numeric blobs and phone-like tails
  s = s.replace(
    /\s+(x?\d{4,}|\d{3}-\d{3,}-\d{3,}|\d{10,}|[0-9]{4,}[A-Z0-9-]*)\s*$/i,
    ""
  ).trim();

  if (!s) return "Unknown";

  // Title-case-ish without being too fancy
  return s
    .toLowerCase()
    .split(/\s+/)
    .map((w) => w.charAt(0).toUpperCase() + w.slice(1))
    .join(" ");
}

/**
 * Normalize a raw merchant/description into:
 * - display name (for UI)
 * - optional kind (p2p, subscription, etc.)
 * - optional categoryHint (e.g. "transfers")
 */
export function normalizeMerchantForDisplayAndCategory(
  raw: string
): NormalizedMerchant {
  if (!raw) {
    return {
      display: "Unknown",
      categoryHint: "unknown",
    };
  }

  // 1) Brand-specific overrides
  const brandRule = MERCHANT_BRAND_RULES.find((rule) => rule.pattern.test(raw));
  if (brandRule) {
    return {
      display: brandRule.normalized,
      kind: brandRule.kind,
      categoryHint: brandRule.categoryHint ?? "other",
    };
  }

  // 2) Generic cleanup
  const display = basicNormalize(raw);

  return {
    display,
    categoryHint: "unknown",
  };
}

/**
 * Minimal shape we need from the backend for merchant charts.
 * (you already have something similar; adjust field names as needed)
 */
export interface MerchantChartRow {
  merchantRaw: string; // backend field merchant / label / description
  spend: number; // positive amount
  txns: number; // transaction count
  category?: string; // optional category slug from backend
}

/**
 * Output row for the Top Merchants chart after normalization + grouping.
 */
export interface MerchantChartRowGrouped {
  merchant: string; // what we show on hover / tooltip
  spend: number;
  txns: number;
  kind?: MerchantKind;
  categoryHint?: MerchantCategoryHint;
  category?: string; // optional category slug from backend
}

/**
 * Normalize merchants and group all Transfers/P2P into a single logical bar
 * "Transfers / P2P". Everyone else keeps their own bucket.
 */
export function normalizeAndGroupMerchantsForChart(
  rows: MerchantChartRow[],
  minSpend = 0.01
): MerchantChartRowGrouped[] {
  const buckets = new Map<string, MerchantChartRowGrouped>();

  for (const row of rows) {
    if (!row) continue;

    const amount = Number(row.spend ?? 0);
    if (!Number.isFinite(amount) || Math.abs(amount) < minSpend) continue;

    const norm = normalizeMerchantForDisplayAndCategory(row.merchantRaw);
    const isTransfers = norm.categoryHint === "transfers";

    const key = isTransfers ? "Transfers / P2P" : norm.display || "Unknown";

    const existing = buckets.get(key) ?? {
      merchant: key,
      spend: 0,
      txns: 0,
      kind: norm.kind,
      categoryHint: norm.categoryHint,
      category: row.category, // preserve category from backend
    };

    existing.spend += amount;
    existing.txns += row.txns ?? 0;

    // Prefer a concrete kind/categoryHint if we didn't have one yet
    if (!existing.kind && norm.kind) existing.kind = norm.kind;
    if (!existing.categoryHint && norm.categoryHint)
      existing.categoryHint = norm.categoryHint;
    // Preserve category from first row if not set
    if (!existing.category && row.category) existing.category = row.category;

    buckets.set(key, existing);
  }

  return Array.from(buckets.values()).sort((a, b) => b.spend - a.spend);
}
